src/magic.h

```diff
@@ -0,0 +1,242 @@
+
+enum energy_type {
+    hp_energy,
+    mana_energy,
+    stamina_energy,
+    bionic_energy,
+    none_energy
+};
+
+enum valid_target {
+    target_ally,
+    target_hostile,
+    target_self,
+    target_ground
+};
+
+class spell_type
+{
+    public:
+        spell_type() = default;
+
+        bool was_loaded = false;
+
+        spell_id id;
+        // spell name
+        std::string name;
+        // spell description
+        std::string description;
+        // spell effect string. used to look up spell function
+        std::string effect;
+        // extra information about spell effect. allows for combinations for effects
+        std::string effect_str;
+
+        // minimum damage this spell can cause
+        int min_damage;
+        // amount of damage increase per spell level
+        float damage_increment;
+        // maximum damage this spell can cause
+        int max_damage;
+
+        // minimum range of a spell
+        int min_range;
+        // amount of range increase per spell level
+        float range_increment;
+        // max range this spell can achieve
+        int max_range;
+
+        // minimum area of effect of a spell (radius)
+        // 0 means the spell only affects the target
+        int min_aoe;
+        // amount of area of effect increase per spell level (radius)
+        float aoe_increment;
+        // max area of effect of a spell (radius)
+        int max_aoe;
+
+        // damage over time deals damage per turn
+
+        // minimum damage over time
+        int min_dot;
+        // increment per spell level
+        float dot_increment;
+        // max damage over time
+        int max_dot;
+
+        // amount of time effect lasts
+
+        // minimum time for effect in moves
+        int min_duration;
+        // increment per spell level in moves
+        // DoT is per turn, but increments can be smaller
+        int duration_increment;
+        // max time for effect in moves
+        int max_duration;
+
+        // amount of damage that is piercing damage
+        // not added to damage stat
+
+        // minimum pierce damage
+        int min_pierce;
+        // increment of pierce damage per spell level
+        float pierce_increment;
+        // max pierce damage
+        int max_pierce;
+
+        // base energy cost of spell
+        int base_energy_cost;
+
+        // spell is restricted to being cast by only this class
+        // if spell_class is empty, spell is unrestricted
+        trait_id spell_class;
+
+        // the difficulty of casting a spell
+        int difficulty;
+
+        // max level this spell can achieve
+        int max_level;
+
+        // base amount of time to cast the spell in moves
+        int base_casting_time;
+
+        // what energy do you use to cast this spell
+        energy_type energy_source;
+
+        damage_type dmg_type;
+
+        // list of valid targets enum
+        std::vector<valid_target> valid_targets;
```

Comment:
Consider making this an `enum_bitset<valid_target>` rather than a `vector`.

Reply:
please advise. i made an attempt, and will try to come back to it later, but i will move on to another portion of this PR first.

Reply:
Yeah, I guess `enum_bitset` should really have some docs, shouldn't it.  But in short, an `enum_bitset<valid_target>` should be usable more-or-less like an array of `bool`s, but indexed by `valid_target` values rather than integers.

The important thing is that you need to add a `last` member to `valid_target` after all the others so it knows how many values there are.